controllers/configurationpolicy_controller.go

```diff
@@ -3279,29 +3286,130 @@ func (r *ConfigurationPolicyReconciler) sendComplianceEvent(instance *policyv1.C
+// template, the default message will be used, appended with details for the error. If no custom
+// template was specified for the current compliance, then the default message is used.
+func (r *ConfigurationPolicyReconciler) customComplianceMessage(plc *policyv1.ConfigurationPolicy) string {
+	customTemplate := plc.Spec.CustomMessage.Compliant
+
+	if plc.Status.ComplianceState != policyv1.Compliant {
+		customTemplate = plc.Spec.CustomMessage.NonCompliant
+	}
+
+	defaultMessage := defaultComplianceMessage(plc)
+
+	// No custom template was provided for the current situation
+	if customTemplate == "" {
+		return defaultMessage
+	}
+
+	customMessage, err := r.doCustomMessage(plc, customTemplate, defaultMessage)
+	if err != nil {
+		return fmt.Sprintf("%v (custom message template failure: %v)", defaultMessage, err.Error())
+	}
+
+	// Add the compliance prefix if not present (it is required by the framework)
+	if !strings.HasPrefix(customMessage, string(plc.Status.ComplianceState)+"; ") {
+		customMessage = string(plc.Status.ComplianceState) + "; " + customMessage
+	}
+
+	return customMessage
+}
+
+// doCustomMessage parses and executes the custom template, returning an error if something goes
+// wrong. The data that the template receives includes the '.DefaultMessage' string and a '.Policy'
+// object, which has the full current state of the configuration policy, including status fields
+// like relatedObjects. If the policy is using the dynamic watcher, then the '.object' field on each
+// related object will have the *full* current state of that object, otherwise only some identifying
+// information is available there.
+func (r *ConfigurationPolicyReconciler) doCustomMessage(
+	plc *policyv1.ConfigurationPolicy, customTemplate string, defaultMessage string,
+) (string, error) {
+	tmpl, err := template.New("custom-msg").Parse(customTemplate)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse custom template: %w", err)
+	}
 
-	if plc.Status.CompliancyDetails == nil || len(plc.Status.CompliancyDetails) == 0 {
-		return result
+	// Converting the policy to a map allows users to access fields via the yaml/json field names
+	// (ie the lowercase versions), which they are likely more familiar with.
+	plcMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(plc)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert policy to unstructured: %w)", err)
 	}
 
-	for _, v := range plc.Status.CompliancyDetails {
-		result += "; "
-		for idx, cond := range v.Conditions {
-			result += cond.Type + " - " + cond.Message
-			if idx != len(v.Conditions)-1 {
-				result += ", "
+	// Only add the full related object information when it can be pulled from the cache
+	if currentlyUsingWatch(plc) {
+		// Paranoid checks to ensure that the policy has a status of the right format
+		plcStatus, ok := plcMap["status"].(map[string]any)
+		if !ok {
+			goto messageTemplating
```

Comment:
Oh goodness. What is this, BASIC programming??? 😆 I had no idea this existed in Go.

Reply:
"any"?

Reply:
It's not Go without `goto` :laughing: 

I'm surprised @mprahl didn't comment on this, I think I've tried to sneak in a goto before... here it's just playing the role of a `break` statement for the `if currentlyUsingWatch` section. Maybe I set the whole thing up poorly, but when I try to re-write it without `goto`, it seems worse to me...
```golang
	// Only add the full related object information when it can be pulled from the cache
	if currentlyUsingWatch(plc) {
		skip := false

		// Paranoid checks to ensure that the policy has a status of the right format
		plcStatus, ok := plcMap["status"].(map[string]any)
		if !ok {
			skip = true
		}

		var relObjs []any

		if !skip {
			relObjs, ok = plcStatus["relatedObjects"].([]any)
			if !ok {
				skip = true
			}
		}

		if !skip {
			for i, relObj := range plc.Status.RelatedObjects {
				objNS := relObj.Object.Metadata.Namespace
				objName := relObj.Object.Metadata.Name
				objGVK := schema.FromAPIVersionAndKind(relObj.Object.APIVersion, relObj.Object.Kind)

				fullObj, err := r.getObjectFromCache(plc, objNS, objName, objGVK)
				if err == nil && fullObj != nil {
					if _, ok := relObjs[i].(map[string]any); ok {
						relObjs[i].(map[string]any)["object"] = fullObj.Object
					}
				}
			}
		}
	}
```

Reply:
These assertions really are just paranoia... since it's coming from a properly typed ConfigurationPolicy, I don't think it's possible they could ever fail. But, if they do fail unchecked, it's a `panic`, so that would be pretty bad.

Reply:
@JustinKuli I decided to let the "goto" slide even though I don't like them in general. It's prevalent in the Go standard library so I let it be an artistic decision. :laughing:

Reply:
I like using goto. Great first step @JustinKuli